### PR TITLE
Feature: Transport address standardization

### PR DESCRIPTION
**Pull request by yashksaini-coder**
Issue: #1093

## What was wrong?

This PR fixes multiple test failures and runtime issues that occurred after the `IRawConnection` and `INetConn` interfaces were updated to include new abstract methods `get_connection_type()` and `get_transport_addresses()`. Additionally, it fixes a critical bug where WebSocket transport connections were failing due to missing background nursery setup.

### Issues Fixed:

1. **WebSocket Transport Background Nursery Missing**
   - **Error**: `OpenConnectionError: No background nursery available. WebSocket transport requires Swarm to set background nursery.`
   - **Impact**: All WebSocket integration tests were failing (8 test failures)
   - **Root Cause**: `Swarm.run()` only set background nursery for QUIC transport, but WebSocket transport also requires it

2. **Missing Abstract Methods in Test Mocks**
   - **Error**: `TypeError: Can't instantiate abstract class ... without an implementation for abstract methods 'get_connection_type', 'get_transport_addresses'`
   - **Impact**: Multiple test suites failing (10+ test failures across different modules)
   - **Root Cause**: Test mock classes (`MockConnection`, `TrioStreamAdapter`, `DummySecuredConn`) didn't implement the newly added abstract methods from `IRawConnection` and `INetConn` interfaces

3. **Type Checker Issues**
   - **Error**: `Missing argument 'cancelled_caught' in function 'trio.CancelScope.__init__'` and `Direction may be uninitialized`
   - **Impact**: Type checker warnings/errors
   - **Root Cause**: Type checker false positives and conditional imports

### 1. WebSocket Background Nursery Setup
- **File**: `libp2p/network/swarm.py`
- **Change**: Added check for transports that have `set_background_nursery` method (including WebSocket) in `Swarm.run()`
- **Approach**: Used `hasattr()` check to avoid circular imports while supporting any transport that needs background nursery
- **Code**:
  ```python
  elif hasattr(self.transport, "set_background_nursery"):
      # WebSocket transport also needs background nursery
      self.transport.set_background_nursery(nursery)
  ```

### 2. Test Mock Updates
- **Files**: 8 test files updated
- **Change**: Added `get_connection_type()` and `get_transport_addresses()` methods to all test mock classes
- **Approach**: 
  - `get_connection_type()` returns `ConnectionType.DIRECT` for test adapters
  - `get_transport_addresses()` returns empty list `[]` for test adapters
  - Fixed `PublicKey`/`PrivateKey` instantiation in `DummySecuredConn` to use `Mock(spec=...)` instead of direct instantiation
- **Files Updated**:
  - `tests/core/network/test_enhanced_swarm.py` - `MockConnection`
  - `tests/core/security/test_insecure_peerstore_integration.py` - `TrioStreamAdapter`
  - `tests/core/relay/test_nat.py` - Mock connections in reachability tests
  - `tests/core/stream_muxer/test_async_context_manager.py` - `DummySecuredConn`
  - `tests/core/stream_muxer/test_mplex_conn.py` - `DummySecuredConn`
  - `tests/core/stream_muxer/test_yamux.py` - `TrioStreamAdapter`
  - `tests/core/stream_muxer/test_yamux_interleaving.py` - `TrioStreamAdapter`
  - `tests/core/stream_muxer/test_yamux_interleaving_EOF.py` - `TrioStreamAdapter`

### 3. Type Checker Fixes
- **File**: `libp2p/network/reconnect_queue.py`
- **Change**: Added `# type: ignore[call-arg]` for `trio.CancelScope()` false positive
- **File**: `libp2p/network/swarm.py`
- **Change**: Moved `Direction` import to top-level to avoid "may be uninitialized" warning

### Testing
- All previously failing tests now pass
- No breaking changes to production code
- All fixes are minimal and targeted

### CODE REVIEW

**Comment by yashksaini-coder** [comment]
@seetadev @acul71 This PR is ready for review, for the setup and progress.

**Comment by seetadev** [comment]
Thanks @yashksaini-coder — this is heading in a really good direction 👍

Appreciate the clear breakdown of what was wrong and how it was fixed. The WebSocket background nursery issue in particular was a real blocker, and the hasattr(self.transport, "set_background_nursery") approach in Swarm.run() is a clean, extensible solution. The updates to the test mocks to align with the new IRawConnection / INetConn abstract methods also look thorough and well-scoped.

Great to see that all the previously failing WebSocket and integration tests are now passing, and that the changes remain minimal and targeted with no production breakage.

There are still a couple of failing core tox checks (3.10 / 3.12). Once those are addressed or confirmed as infra-related, I think we’ll be in good shape to move forward.

Looping in @pacrob for an additional review, especially on the Swarm and transport-side changes.

Thanks again for the solid work and for driving this forward 🚀

**Comment by yashksaini-coder** [comment]
> Thanks @yashksaini-coder — this is heading in a really good direction 👍
> 
> Appreciate the clear breakdown of what was wrong and how it was fixed. The WebSocket background nursery issue in particular was a real blocker, and the hasattr(self.transport, "set_background_nursery") approach in Swarm.run() is a clean, extensible solution. The updates to the test mocks to align with the new IRawConnection / INetConn abstract methods also look thorough and well-scoped.
> 
> Great to see that all the previously failing WebSocket and integration tests are now passing, and that the changes remain minimal and targeted with no production breakage.
> 
> There are still a couple of failing core tox checks (3.10 / 3.12). Once those are addressed or confirmed as infra-related, I think we’ll be in good shape to move forward.
> 
> Looping in @pacrob for an additional review, especially on the Swarm and transport-side changes.
> 
> Thanks again for the solid work and for driving this forward 🚀

@seetadev The Failing ci-checks are infra related issue, I will raise a Discussion highlighting the main core issue and also attach failing PRs because of that to address PRs that are correct and are only showing these particular infra-related ci checks

**Comment by acul71** [comment]
Hi @yashksaini-coder and @seetadev 
Hoping that there is no external code didn't re-implemented custom core modules (about breaking changes), the internal cases are covered.
Take a look at Major and minor issues

#########################################################


# AI Pull Request Review: PR #1127

**Review Date:** 2026-01-28  
**PR Title:** Feature: Transport address standardization  
**Author:** yashksaini-coder  
**Issue:** #1093

---

## 1. Summary of Changes

This PR implements **Phase 1** of the Transport Address Standardization Architecture Refactoring described in issue #1093. The changes add standardized methods to connection interfaces to expose transport address information and connection type without requiring access to low-level `raw_conn` properties.

### Issues Addressed

- **Issue #1093**: Transport Address Standardization Architecture Refactoring
  - Problem: Code needs to access low-level `raw_conn` to determine connection type (direct vs relayed) and actual transport addresses, requiring `type: ignore` comments and breaking encapsulation
  - Solution: Adds `get_transport_addresses()` and `get_connection_type()` methods to connection interfaces (`IRawConnection`, `ISecureConn`, `IMuxedConn`, `INetConn`)

### Modules Affected

- **Core Interfaces** (`libp2p/abc.py`): Added abstract methods to `IRawConnection`, `ISecureConn`, `IMuxedConn`, and `INetConn`
- **Connection Types** (`libp2p/connection_types.py`): New file defining `ConnectionType` enum
- **Connection Implementations**:
  - `libp2p/network/connection/raw_connection.py`: Implements new methods
  - `libp2p/network/connection/swarm_connection.py`: Stores and returns actual transport addresses
  - `libp2p/stream_muxer/mplex/mplex.py`: Delegates to secured_conn
  - `libp2p/stream_muxer/yamux/yamux.py`: Delegates to secured_conn
  - `libp2p/transport/quic/connection.py`: Implements methods for QUIC
- **Security Layer**:
  - `libp2p/security/insecure/transport.py`: Delegates to underlying connection
  - `libp2p/security/secure_session.py`: Delegates to underlying connection
- **Relay Protocol**:
  - `libp2p/relay/circuit_v2/nat.py`: Uses `get_connection_type()` instead of checking addresses
  - `libp2p/relay/circuit_v2/protocol.py`: Sets connection type when creating relay connections
  - `libp2p/relay/circuit_v2/transport.py`: Returns `RELAYED` connection type
- **Network Layer** (`libp2p/network/swarm.py`): Captures transport info during connection establishment
- **Test Files**: Updated 8 test files to implement new abstract methods in mock classes

### Breaking Changes

⚠️ **This PR introduces breaking changes to internal connection interfaces.**

#### Interface Changes

Four core connection interfaces now require implementation of two new abstract methods:

1. **`IRawConnection`** (lines 87-102 in `libp2p/abc.py`)
   - Added: `get_transport_addresses() -> list[Multiaddr]`
   - Added: `get_connection_type() -> ConnectionType`

2. **`ISecureConn`** (inherits from `IRawConnection`)
   - Inherits both abstract methods from `IRawConnection`
   - Implementations must provide these methods

3. **`IMuxedConn`** (lines 251-262 in `libp2p/abc.py`)
   - Added: `get_transport_addresses() -> list[Multiaddr]`
   - Added: `get_connection_type() -> ConnectionType`

4. **`INetConn`** (lines 406-417 in `libp2p/abc.py`)
   - Added: `get_connection_type() -> ConnectionType`
   - `get_transport_addresses()` already existed but is now abstract

#### Impact Assessment

**Internal Implementations (✅ All Updated):**
- `RawConnection` (IRawConnection) - ✅ Implemented
- `QUICConnection` (IRawConnection, IMuxedConn) - ✅ Implemented
- `SwarmConn` (INetConn) - ✅ Implemented
- `Yamux` (IMuxedConn) - ✅ Implemented
- `Mplex` (IMuxedConn) - ✅ Implemented
- `TrackedRawConnection` (IRawConnection) - ✅ Implemented
- `BaseSession` (ISecureConn) - ✅ Implemented (via inheritance)
- All test mocks - ✅ Updated

**Note on External Impact:**
- These interfaces are **not exported** in the main `libp2p/__init__.py` public API
- All implementations found in the codebase are internal to the `libp2p` package
- These interfaces are primarily used as **type hints** in public APIs
- **No external implementations found** - impact appears to be limited to internal code only

---

## 2. Branch Sync Status and Merge Conflicts

### Branch Sync Status

- **Status:** ✅ **In sync** with origin/main
- **Details:** Branch is **0 commits behind** and **11 commits ahead** of origin/main
  - The branch contains 11 commits not yet in main
  - Branch is fully up to date with main (no missing commits)

### Merge Conflict Analysis

- **Conflicts Detected:** ✅ **No conflicts** - PR can be merged cleanly
- **Status:** The test merge completed successfully with no conflicts detected
- **Merge Readiness:** ✅ Ready for merge (after addressing other review items)

---

## 3. Strengths

1. **Well-Structured Interface Design**: The addition of `get_transport_addresses()` and `get_connection_type()` methods follows a clean delegation pattern through the connection stack (Raw → Secure → Muxed → Network)

2. **Comprehensive Implementation**: All connection types (RawConnection, QUICConnection, Mplex, Yamux, security layers) have been updated to implement the new methods

3. **Proper Abstraction**: The `ConnectionType` enum provides a clean, type-safe way to determine connection type without string parsing or address inspection

4. **Test Coverage**: All test mocks have been updated to implement the new abstract methods, ensuring tests continue to pass

5. **Backward Compatibility Consideration**: The `SwarmConn.set_transport_info()` method allows setting transport information during connection establishment, with fallback to peerstore addresses if not set

6. **Error Handling**: Graceful error handling in `swarm.py` when setting transport info (logs debug message on failure rather than crashing)

7. **Documentation**: Good docstrings explaining the purpose of new methods and what they return

8. **Refactoring in NAT Module**: The NAT traversal code now uses `get_connection_type()` instead of fragile string checking for `/p2p-circuit`, making the code more maintainable

---

## 4. Issues Found

### Critical

None identified.

### Major

1. **Duplicate `ConnectionType` Enum Definition**
   - **File:** `libp2p/custom_types.py`
   - **Line(s):** 46-55
   - **Issue:** `ConnectionType` is defined in both `libp2p/connection_types.py` (new file) and `libp2p/custom_types.py`. This creates a duplicate definition that could lead to import confusion and type inconsistencies.
   - **Impact:** 
     - Two separate enum classes with the same name but different identities
     - Type checking may fail if different files import from different locations
     - Runtime behavior may be unpredictable if enum instances are compared across imports
   - **Suggestion:** Remove the `ConnectionType` enum from `libp2p/custom_types.py` since it's already defined in `libp2p/connection_types.py`. All imports should use `from libp2p.connection_types import ConnectionType` or `from libp2p.abc import ConnectionType` (which re-exports it).

2. **Missing Newsfragment** (See Section 7 for details)
   - **Severity:** CRITICAL / BLOCKER
   - **Issue:** No newsfragment file exists for issue #1093
   - **Impact:** PR cannot be approved without a newsfragment
   - **Action Required:** Add `newsfragments/1093.feature.rst` (or appropriate type)

### Minor

1. **Inconsistent Import Pattern**
   - **File:** `libp2p/abc.py`
   - **Line(s):** 27-29, 35-40
   - **Issue:** `ConnectionType` is imported from `libp2p.connection_types`, but other types are imported from `libp2p.custom_types`. While this works, it's inconsistent. Since `ConnectionType` is now in `connection_types.py`, consider whether it should be moved to `custom_types.py` for consistency, or if the separation is intentional.
   - **Suggestion:** If keeping separate files, document the rationale. If consolidating, move `ConnectionType` to `custom_types.py` and remove `connection_types.py`.

2. **Type Annotation in `raw_connection.py`**
   - **File:** `libp2p/network/connection/raw_connection.py`
   - **Line(s):** 193-200
   - **Issue:** The `get_transport_addresses()` method creates a Multiaddr by assuming TCP protocol (`/ip4/{ip}/tcp/{port}`) with a comment "Assume TCP for now, but this should be more sophisticated". This is a hardcoded assumption that may not work for all transport types.
   - **Suggestion:** Either:
     - Extract protocol information from the underlying stream if available
     - Make the protocol configurable via constructor parameter
     - Document this limitation clearly in the docstring

3. **Redundant `hasattr` Check** (Needs Verification)
   - **File:** `libp2p/io/abc.py`
   - **Line(s):** 84, 94, 104
   - **Issue:** The code checks `hasattr(self, "conn")` after already checking `self.conn is not None`. Since `conn` is a class attribute that's always initialized in `__init__`, and accessing `self.conn` in the `is not None` check would already raise `AttributeError` if it didn't exist, the `hasattr` check appears redundant.
   - **Note:** While the check appears redundant, it may have been added defensively. Removing it should be safe since:
     - `conn` is a class attribute (`conn: Any | None`) that's always set in `__init__`
     - The `self.conn is not None` check already ensures the attribute exists (or would raise `AttributeError`)
   - **Suggestion:** If removing, verify with linting and type checking:
     ```python
     if self.conn is not None and hasattr(self.conn, "get_transport_addresses"):
         return self.conn.get_transport_addresses()
     ```
   - **Verification Required:** Test that removing `hasattr(self, "conn")` doesn't cause mypy/pyright errors or break any edge cases.

4. **Missing Type Import in `raw_connection.py`**
   - **File:** `libp2p/network/connection/raw_connection.py`
   - **Line(s):** Import section
   - **Issue:** `ConnectionType` is imported from `libp2p.abc`, but it's actually defined in `libp2p.connection_types`. While `libp2p.abc` re-exports it, it would be clearer to import directly from the source.
   - **Suggestion:** Consider importing directly: `from libp2p.connection_types import ConnectionType` for clarity, or ensure `libp2p.abc` properly re-exports it in `__all__`.

5. **Incomplete Error Handling in `swarm.py`**
   - **File:** `libp2p/network/swarm.py`
   - **Line(s):** 302-311
   - **Issue:** The exception handler catches all exceptions (`except Exception as e`) when setting transport info. While this prevents crashes, it silently swallows all errors, which could hide real issues.
   - **Suggestion:** Consider being more specific about which exceptions to catch, or at least log at a higher level (warning) for unexpected exceptions while keeping debug level for expected cases.

---

## 5. Security Review

No security vulnerabilities identified. The changes are focused on interface additions and information exposure, which are read-only operations that don't introduce new attack surfaces.

**Considerations:**
- The new methods expose transport address information, which could potentially leak network topology information. However, this information was already accessible through `raw_conn` (with type ignores), so this change doesn't introduce new information disclosure risks.
- The connection type information is useful for security-sensitive code (e.g., NAT traversal, relay validation) and helps improve security by providing a standardized way to check connection types.

**Risk Assessment:**
- **Risk:** None identified
- **Impact:** N/A
- **Mitigation:** N/A

---

## 6. Documentation and Examples

### Documentation Quality

- ✅ **Interface Docstrings**: All new abstract methods have clear docstrings explaining their purpose
- ✅ **Implementation Docstrings**: Concrete implementations document their behavior
- ✅ **Enum Documentation**: `ConnectionType` enum has a docstring explaining its purpose

### Missing Documentation

1. **Module-Level Documentation**: `libp2p/connection_types.py` is a new module but lacks a module docstring explaining its purpose and relationship to the architecture
2. **Migration Guide**: No documentation for external code that implements these interfaces on how to add the new methods
3. **API Documentation**: The new methods should be added to the API reference documentation

### Recommendations

- Add a module docstring to `libp2p/connection_types.py` explaining that this module provides connection type enumerations for the transport address standardization architecture
- Consider adding a migration note in the release notes (via newsfragment) for developers implementing these interfaces
- Update API documentation to include the new methods in the interface documentation

---

## 7. Newsfragment Requirement

⚠️ **CRITICAL: Missing Newsfragment - BLOCKER**

### Status

- ❌ **Newsfragment Missing**: No newsfragment file exists for issue #1093
- ✅ **Issue Reference**: PR correctly references issue #1093 in the description

### Required Action

**File to Create:** `newsfragments/1093.feature.rst`

**Content Requirements:**
- **Type:** `.feature.rst` (new feature/interface addition)
- **Content:** User-facing description of the change (not implementation details)
- **Format:** ReST-formatted, must end with a newline
- **Focus:** Describe what users can now do, not how it was implemented

**Suggested Content:**
```
Added standardized methods to connection interfaces to access transport address information and connection type without requiring low-level connection access. Connections now expose ``get_transport_addresses()`` and ``get_connection_type()`` methods to determine if a connection is direct or relayed, and to retrieve the actual transport addresses used.
```

**Action Required:** PR approval is **BLOCKED** until a newsfragment file is added.

---

## 8. Tests and Validation

### Linting (`make lint`)

- **Status:** ✅ **PASSED**
- **Exit Code:** 0
- **Results:** All linting checks passed:
  - ✅ check yaml
  - ✅ check toml
  - ✅ fix end of files
  - ✅ trim trailing whitespace
  - ✅ pyupgrade
  - ✅ ruff (legacy alias)
  - ✅ ruff format
  - ✅ mdformat
  - ✅ mypy with all dev dependencies present
  - ✅ pyrefly typecheck locally
  - ✅ Check for .rst files in the top-level directory

**No errors, warnings, or issues reported.**

### Type Checking (`make typecheck`)

- **Status:** ✅ **PASSED**
- **Exit Code:** 0
- **Results:** 
  - ✅ mypy with all dev dependencies present
  - ✅ pyrefly typecheck locally

**No type errors, warnings, or issues reported.**

### Test Execution (`make test`)

- **Status:** ✅ **PASSED**
- **Exit Code:** 0
- **Test Summary:**
  - **Total Tests:** 1920 items
  - **Passed:** 1905
  - **Skipped:** 15
  - **Failed:** 0
  - **Errored:** 0
  - **Warnings:** 25 (deprecation and resource warnings, not related to PR changes)
  - **Execution Time:** 93.29 seconds

**All tests pass successfully.** The 25 warnings are pre-existing deprecation and resource warnings unrelated to this PR.

### Documentation Build (`make linux-docs`)

- **Status:** ✅ **PASSED**
- **Exit Code:** 0
- **Results:**
  - ✅ Sphinx HTML build succeeded
  - ✅ Doctest execution: 109 tests, 0 failures
  - ✅ Newsfragment validation passed (for existing fragments)

**No documentation build errors or warnings.**

---

## 9. Recommendations for Improvement

1. **Remove Duplicate Enum** (HIGH PRIORITY)
   - Remove `ConnectionType` from `libp2p/custom_types.py`
   - Ensure all imports use `libp2p.connection_types` or `libp2p.abc` (which re-exports it)
   - Verify no code imports `ConnectionType` from `custom_types`

2. **Add Newsfragment** (CRITICAL - BLOCKER)
   - Create `newsfragments/1093.feature.rst` with user-facing description
   - Ensure file ends with a newline character

3. **Improve Transport Protocol Detection**
   - Enhance `RawConnection.get_transport_addresses()` to detect transport protocol (TCP/UDP/QUIC) from the underlying stream rather than hardcoding TCP
   - Consider adding protocol information to the connection metadata

4. **Code Cleanup**
   - Remove redundant `hasattr(self, "conn")` checks in `libp2p/io/abc.py`
   - Consider more specific exception handling in `swarm.py` transport info setting

5. **Documentation Enhancements**
   - Add module docstring to `libp2p/connection_types.py`
   - Update API documentation to include new interface methods
   - Consider adding migration notes for external implementers

6. **Consistency Review**
   - Review import patterns: should `ConnectionType` be in `custom_types.py` for consistency with other type definitions, or is the separation intentional?
   - Document the rationale for having a separate `connection_types.py` module

---

## 10. Questions for the Author

1. **Duplicate Enum**: Why is `ConnectionType` defined in both `connection_types.py` and `custom_types.py`? Should one be removed, or is there a specific reason for the duplication?

2. **Module Organization**: What's the rationale for creating a separate `connection_types.py` module instead of adding `ConnectionType` to `custom_types.py` where other type definitions live?

3. **Transport Protocol Assumption**: In `RawConnection.get_transport_addresses()`, the code assumes TCP protocol. Is there a plan to make this more sophisticated, or is TCP the only transport type that will use `RawConnection`?

4. **Error Handling Strategy**: In `swarm.py`, all exceptions are caught when setting transport info. Are there specific exceptions that should be handled differently, or should some exceptions be allowed to propagate?

5. **Backward Compatibility**: For external code that implements `IRawConnection`, `ISecureConn`, or `IMuxedConn`, will this be a breaking change requiring immediate updates, or is there a deprecation period planned?

6. **Future Phases**: This PR implements Phase 1 of the architecture refactoring. Are there plans for when Phase 2 (storing connection metadata during establishment) and Phase 3+ will be implemented?

---

## 11. Overall Assessment

### Quality Rating: **Good** → **Needs Work**

The PR implements a solid foundation for transport address standardization, but has a critical blocker (missing newsfragment) and a major issue (duplicate enum definition) that must be addressed before approval.

### Security Impact: **None**

No security vulnerabilities introduced. The changes improve code quality by removing the need for `type: ignore` comments and providing a standardized interface.

### Merge Readiness: **Needs Fixes**

**Blockers:**
- ❌ Missing newsfragment for issue #1093 (CRITICAL)
- ⚠️ Duplicate `ConnectionType` enum definition (MAJOR - must be resolved)

**Required Before Merge:**
1. Add `newsfragments/1093.feature.rst` file
2. Remove duplicate `ConnectionType` enum from `custom_types.py`
3. Verify all imports are consistent

**Recommended Before Merge:**
1. Address minor code quality issues (redundant checks, error handling)
2. Add module documentation
3. Consider transport protocol detection improvements

### Confidence: **High**

The implementation is well-structured and follows good patterns. The test suite passes completely, and the changes align with the architecture described in issue #1093. Once the blockers are addressed, this PR will be ready for merge.

---

## Summary

This PR successfully implements Phase 1 of the Transport Address Standardization Architecture by adding standardized methods to connection interfaces. The implementation is comprehensive, well-tested, and follows good design patterns. However, **two critical issues must be resolved**:

1. **Missing newsfragment** - A blocker that prevents PR approval
2. **Duplicate enum definition** - Must be resolved to prevent type inconsistencies

Once these are addressed, along with the recommended improvements, this PR will be ready for approval and merge.

**Comment by yashksaini-coder** [comment]
## What I Fixed in This PR
- Issue: #1093

### Critical Issues
- **Missing changelog file**: Added the newsfragment file that describes these changes
- **Duplicate code**: Removed a duplicate `ConnectionType` enum that was defined in two places
- **Unnecessary checks**: Cleaned up redundant code checks in the I/O module
- **IPv6 support**: Fixed hardcoded TCP protocol handling to work with IPv6 addresses
- **Better error logging**: Made error messages more helpful - now only warns about unexpected errors instead of logging everything

## Design Choices

**Why keep connection types in a separate file?**
It follows the pattern used in other libp2p implementations (Go, JS, Rust). Keeps networking code organized.

**Why default to TCP?**
Most basic connections use TCP. Other protocols like QUIC and WebSocket have their own implementations.

**Will this break existing code?**
No, I only changed internal interfaces. External users won't be affected, which is why this is a minor version bump.

## What's Done and What's Coming

**In this PR:**
- Better connection address handling
- Clearer error messages
- IPv6 support

**Future work (2026):**
- Improve relay and NAT address handling (Q2)
- Add address filtering and validation (Q3)
- Update discovery and routing systems (Q3-Q4)

@acul71 Thanks for the detailed review! I've addressed all the issues you found. Also the Phase 2 is almost 60% completed, within this PR itself. What I planned was complete and present a phase and gather community feedback on the next further areas

For detailed updates and progress please refer to this discussion page, https://github.com/libp2p/py-libp2p/discussions/1177

File: libp2p/relay/circuit_v2/transport.py [hunk 1]
```diff
@@ -581,12 +601,11 @@ async def _dial_via_circuit_addr(
...
-            raw_conn = RawConnection(stream=relay_stream, initiator=True)
-            return TrackedRawConnection(
-                wrapped=raw_conn,
-                relay_id=relay_peer_id,
-                tracker=self.performance_tracker,
+            return RawConnection(
+                stream=relay_stream,
+                initiator=True,
+                connection_type=ConnectionType.RELAYED,
+                addresses=[circuit_ma],
```
Comment (pacrob) [inline libp2p/relay/circuit_v2/transport.py]: This drops the `Tracked` portion, but above in `dial_peer_info` (~line 358), there's a comment saying a `TrackedRawConnection` is expected. How import is the tracked aspect? Why is it being dropped here?

Comment (yashksaini-coder) [reply libp2p/relay/circuit_v2/transport.py]: @pacrob Great highlight, I've made both connection paths track circuits properly - now connections are counted correctly when they open AND close, preventing memory leaks and securing accurate load balancing.

File: libp2p/relay/circuit_v2/protocol.py [hunk 2]
```diff
@@ -546,8 +549,14 @@ async def handle_incoming_connection(
         """
         try:
             # Create raw connection
-            raw_conn = RawConnection(stream=stream, initiator=False)
-            ma = multiaddr.Multiaddr(remote_peer_id)
+            peer_id = ID(remote_peer_id)
+            ma = multiaddr.Multiaddr(f"/p2p/{peer_id.to_base58()}")
```
Comment (pacrob) [inline libp2p/relay/circuit_v2/protocol.py]: Relay connections should include `p2p-circuit`, should they not?

File: libp2p/relay/circuit_v2/transport.py
Comment (pacrob) [inline libp2p/relay/circuit_v2/transport.py:1]: At lines 461 and 1050 in this file we are creating `RawConnection`s with the default `ConnectionType.DIRECT`. Since these are also handling relayed connections, should we update those parameters in this PR too?

**Comment by pacrob** [comment]
Consider adding testing around relay connections reporting `ConnectionType.RELAYED` and that their maddrs include `/p2p-circuit`.

Thank you!

File: libp2p/relay/circuit_v2/protocol.py (hunk 2 above)
Comment (yashksaini-coder) [reply libp2p/relay/circuit_v2/protocol.py]: Relay connections now include the `p2p-circuit` protocol to indicate they're relayed connections.

File: libp2p/relay/circuit_v2/transport.py
Comment (yashksaini-coder) [reply libp2p/relay/circuit_v2/transport.py:1]: Both RawConnection instances now properly use ConnectionType.RELAYED with the correct circuit multiaddr format /p2p/{relay_id}/p2p-circuit/p2p/{peer_id}.

**Comment by yashksaini-coder** [comment]
> Consider adding testing around relay connections reporting `ConnectionType.RELAYED` and that their maddrs include `/p2p-circuit`.
> 
> Thank you!

Add test cases and reviewed the comments on the addressed areas

File: libp2p/relay/circuit_v2/transport.py
Comment (pacrob) [inline libp2p/relay/circuit_v2/transport.py]: It's not part of your changes, but this docstring is incorrect, can be removed. Thanks!

**Review by pacrob — APPROVED** [review]
One small request for a docstring correction, otherwise good to merge!

File: libp2p/relay/circuit_v2/transport.py
Comment (yashksaini-coder) [reply libp2p/relay/circuit_v2/transport.py]: Thank you sm paul, the docstring change has been removed, can we do final review and merge @pacrob @seetadev

**Comment by yashksaini-coder** [comment]
> One small request for a docstring correction, otherwise good to merge!

Hi @pacrob the docstring has been corrected, and glad to see it is ready to merge

**Comment by seetadev** [comment]
@yashksaini-coder : Thanks Yash. Appreciate it.

Looks like the PR is ready for final review + merge. Will merge it tonight after one more review.

Please open a discussion page and share an example on how you would like to demonstrate transport address standardization via a use case. Helpful for new contributors to learn and contribute.
